sdks/python/apache_beam/ml/transforms/handlers.py

```diff
@@ -0,0 +1,436 @@
+        transformed_types[name] = typing.Sequence[bytes]
+      else:
+        # TODO: This else condition won't be hit since TFT doesn't output
+        # other than float, int and bytes. Refactor the code here.
+        raise RuntimeError(
+            'Unsupported feature type: %s encountered' % feature_type)
+    logging.info(transformed_types)
+    return transformed_types
+
+  def process_data(
+      self, raw_data: beam.PCollection[tft_process_handler_input_type]
+  ) -> beam.PCollection[typing.Union[
+      beam.Row, Dict[str, np.ndarray], pa.RecordBatch]]:
+    """
+    This method also computes the required dataset metadata for the tft
+    AnalyzeDataset/TransformDataset step.
+
+    This method uses tensorflow_transform's Analyze step to produce the
+    artifacts and Transform step to apply the transforms on the data.
+    Artifacts are only produced if the artifact_mode is set to `produce`.
+    If artifact_mode is set to `consume`, then the artifacts are read from the
+    artifact_location, which was previously used to store the produced
+    artifacts.
+    """
+    if self.artifact_mode == ArtifactMode.PRODUCE:
+      # If we are computing artifacts, we should fail for windows other than
+      # default windowing since for example, for a fixed window, each window can
+      # be treated as a separate dataset and we might need to compute artifacts
+      # for each window. This is not supported yet.
+      self._fail_on_non_default_windowing(raw_data)
+      element_type = raw_data.element_type
+      column_type_mapping = {}
+      if (isinstance(element_type, RowTypeConstraint) or
+          native_type_compatibility.match_is_named_tuple(element_type)):
+        column_type_mapping = self._map_column_names_to_types(
+            row_type=element_type)
+        # convert Row or NamedTuple to Dict
+        raw_data = (
+            raw_data
+            | ConvertNamedTupleToDict().with_output_types(
+                Dict[str, typing.Union[tuple(column_type_mapping.values())]]))
+        # AnalyzeAndTransformDataset raise type hint since this is
+        # schema'd PCollection and the current output type would be a
+        # custom type(NamedTuple) or a beam.Row type.
+      else:
+        column_type_mapping = self._map_column_names_to_types_from_transforms()
+      raw_data_metadata = self.get_raw_data_metadata(
+          input_types=column_type_mapping)
+      # Write untransformed metadata to a file so that it can be re-used
+      # during Transform step.
+      metadata_io.write_metadata(
+          metadata=raw_data_metadata,
+          path=os.path.join(self.artifact_location, RAW_DATA_METADATA_DIR))
+    else:
+      # Read the metadata from the artifact_location.
+      if not os.path.exists(os.path.join(
+          self.artifact_location, RAW_DATA_METADATA_DIR, SCHEMA_FILE)):
+        raise FileNotFoundError(
+            "Raw data metadata not found at %s" %
```

Comment:
We should be more descriptive in these errors (something along the lines of "you're running in consume mode" what that means, and "have you run this in produce mode?")

Reply:
Changed wording